src/client.rs

```diff
@@ -904,36 +904,102 @@ pub struct AudioHandler {
 }
 
 #[cfg(not(any(target_os = "android", target_os = "linux")))]
-struct AudioBuffer(pub Arc<std::sync::Mutex<ringbuf::HeapRb<f32>>>);
+struct AudioBuffer(
+    pub Arc<std::sync::Mutex<ringbuf::HeapRb<f32>>>,
+    usize,
+    [usize; 30],
+);
 
 #[cfg(not(any(target_os = "android", target_os = "linux")))]
 impl Default for AudioBuffer {
     fn default() -> Self {
-        Self(Arc::new(std::sync::Mutex::new(
-            ringbuf::HeapRb::<f32>::new(48000 * 2 * AUDIO_BUFFER_MS / 1000), // 48000hz, 2 channel
-        )))
+        Self(
+            Arc::new(std::sync::Mutex::new(
+                ringbuf::HeapRb::<f32>::new(48000 * 2 * AUDIO_BUFFER_MS / 1000), // 48000hz, 2 channel
+            )),
+            48000 * 2,
+            [0; 30],
+        )
     }
 }
 
 #[cfg(not(any(target_os = "android", target_os = "linux")))]
 impl AudioBuffer {
-    pub fn resize(&self, sample_rate: usize, channels: usize) {
+    pub fn resize(&mut self, sample_rate: usize, channels: usize) {
         let capacity = sample_rate * channels * AUDIO_BUFFER_MS / 1000;
         let old_capacity = self.0.lock().unwrap().capacity();
         if capacity != old_capacity {
             *self.0.lock().unwrap() = ringbuf::HeapRb::<f32>::new(capacity);
+            self.1 = sample_rate * channels;
             log::info!("Audio buffer resized from {old_capacity} to {capacity}");
         }
     }
 
-    // clear when full to avoid long time noise
-    #[inline]
-    pub fn clear_if_full(&self) {
-        let full = self.0.lock().unwrap().is_full();
-        if full {
-            self.0.lock().unwrap().clear();
-            log::trace!("Audio buffer cleared");
+    fn try_shrink(&mut self, having: usize) {
+        extern crate chrono;
+        use chrono::prelude::*;
+
+        static mut tms: i64 = 0;
+        let dt = Local::now().timestamp_millis();
+        unsafe {
+            if tms == 0 {
+                tms = dt;
+                return;
+            } else if dt < tms + 12000 {
+                return;
+            }
+            tms = dt;
+        }
+
+        let mut i = (having * 10) / self.1;
```

Comment:
Do you want to count whether the buffer is always in a high-capacity state? If it is, maybe should put it at the top of  the function.

Reply:
yes, I mdee this change jsut before the final commit to make the code more beautiful, without considering it carefully

Reply:
> And how about another solution: the underrun bytes length are the real accumelated bytes length, skip it when it reaches a specific threshold.

It maybe be good in therory, but considering the jitter in audio streams, this can lead to more data insufficiency to from the device pulling view and more packet discard at audio pushing site.